toolchain/check/impl_lookup.cpp

```diff
@@ -130,6 +131,40 @@ auto LookupImplWitness(Context& context, SemIR::LocId loc_id,
     }
   }
 
+  auto& stack = context.impl_lookup_stack();
+  // Deduction of the interface parameters can do further impl lookups, and we
+  // need to ensure we terminate.
+  //
+  // https://docs.carbon-lang.dev/docs/design/generics/details.html#acyclic-rule
+  // - We look for violations of the acyclic rule by seeing if a previous lookup
+  //   had all the same type inputs.
+  // - The `interface_const_id` encodes the entire facet type being looked up,
+  //   including any specific parameters for a generic interface.
+  //
+  // TODO: Implement the termination rule, which requires looking at the
+  // complexity of the types on the top of (or throughout?) the stack:
+  // https://docs.carbon-lang.dev/docs/design/generics/details.html#termination-rule
+  for (auto entry : stack) {
+    if (entry.type_const_id == type_const_id &&
+        entry.interface_const_id == interface_const_id) {
+      CARBON_DIAGNOSTIC(ImplLookupCycle, Error,
+                        "cycle found in lookup of interface {0} for type {1}",
+                        std::string, SemIR::TypeId);
+      context.emitter()
+          .Build(loc_id, ImplLookupCycle, "<TODO: interface name>",
+                 // FIXME: Invalid for facet values??? Write a test where cycle
```

Comment:
Minor, but we typically use TODO for things that are merged, FIXME is more of a "fix before merging"; I'm not sure which way you're heading here

Reply:
Done